src/coordtest/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+                                            }
+                                        },
+                                        _ => panic!("expected SendingRows"),
+                                    };
+                                }
+                            }
+                            Err(err) => {
+                                failed = Err(err);
+                            }
+                        };
+                        match failed {
+                            Ok(_) => {
+                                break;
+                            }
+                            Err(err) => {
+                                if start.elapsed() > Duration::from_secs(5) {
+                                    panic!("{}", err);
+                                }
+                                // Bump the timestamp. This is necessary because sources ingest at varying
+                                // rates and we need to allow sinces to move forward so we can see new data.
+                                let mut ts = ct.timestamp.lock().unwrap();
+                                *ts += 1;
+                            }
+                        }
+                    }
+                    "".into()
+                }
+                "update-upper" => {
+                    let catalog = ct.make_catalog().await;
+                    let mut updates = vec![];
+                    for line in tc.input.lines() {
+                        let mut line = line.split_whitespace();
+                        let id = catalog.get(line.next().unwrap());
+                        let ts: Timestamp = line.next().unwrap().parse().unwrap();
+                        assert!(line.next().is_none());
+                        // A ts <= 1 won't do anything.
+                        assert!(ts > 1);
+                        let upper = ct.uppers.entry(id).or_insert(0);
+                        let mut batch: ChangeBatch<Timestamp> = ChangeBatch::new_from(*upper, -1);
+                        *upper = ts;
+                        batch.update(ts, 1);
+                        updates.push((id, batch));
+                    }
+                    ct.inner_feedback_tx
+                        .send(WorkerFeedbackWithMeta {
+                            worker_id: 0,
+                            message: WorkerFeedback::FrontierUppers(updates),
+                        })
+                        .unwrap();
+                    "".into()
+                }
+                "inc-timestamp" => {
+                    let inc: u64 = tc.input.trim().parse().unwrap();
+                    let mut ts = ct.timestamp.lock().unwrap();
+                    *ts += inc;
+                    "".into()
+                }
+                "create-file" => {
+                    let path = ct.temp_dir.path().join(tc.args["name"][0].clone());
+                    let mut f = File::create(path).unwrap();
```

Comment:
nit: this truncates if the file already exists idk how much we care about that

Reply:
Yes, the docs for `create-file` say "creates and truncates" so this is desired.